### PR TITLE
chore: bump minimum supported Python version to 3.11

### PR DESCRIPTION
**Pull request by bowenliang123**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [x] Please open an issue before creating a PR or link to an existing issue  #10399
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

- drop the support for Python 3.10, support and tested with Python 3.11 and 3.12
- rerunning `dev/reformat`
  - to fix [UP042](https://docs.astral.sh/ruff/rules/replace-str-enum/#replace-str-enum-up042) linter rule violations, auto fixed by  with `--unsafe-fixes`
  - other linter auto fixes
- updated poetry lockfile, stripping dependency details for Python 3.10

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: api/README.md [hunk 1]
```diff
@@ -3,7 +3,6 @@
 ## Usage
 
 > [!IMPORTANT]
```
Comment (laipz8200) [inline api/README.md]: This line should also be removed.

**Review by laipz8200 — COMMENTED** [review]
Thank you for doing this great work.

IMO, the updates for GitHub Actions, documentation, and code should be separated into three PRs. Would you mind making this change?

There’s no need to rush, as we still need to test the new Python version.

**Comment by bowenliang123** [comment]
Thanks for the review. Python 3.11 brings performance improvements and many helpful features and typing.

As for your suggestion of separated PR, it seems that we still have to put them in one.
Once we raise the required minimum version in `pyproject.toml` project specification, it will immediately fail on Python 3.10 including:
- failed to install Python dependency via Poetry with Python version checks
- failed to pass Poetry's lockfile checks
- failed to pass Ruff linter checks, as Ruff runs the checking on different baseline on based minimum Python versions, especially for the [UP](https://docs.astral.sh/ruff/rules/#pyupgrade-up) pyupgrade rules

**Comment by cuckootan** [comment]
I think it's better to keep support python3.10, due to the great cost of upgrading docker image(and other parts).

**Comment by laipz8200** [comment]
@cuckootan We have to do this upgrade cause many new libraries require the new Python version. You can start preparing to do this upgrade from now on.

**Comment by bowenliang123** [comment]
Marking this PR to draft temporarily, as api tests failed to pass tests in `test_iteration.py` without no clear reasons.

**Review by laipz8200 — APPROVED** [review]
LGTM

**Pull request merged**